components/filesystem/include/Filesystem.hpp

```diff
@@ -0,0 +1,129 @@
+#ifndef FILESYSTEM_HPP
+#define FILESYSTEM_HPP
+
+#include <sys/_default_fcntl.h>
+
+#include <string>
+#include <fstream>
+
+#include "Component.hpp"
+#include "esp_err.h"
+#include "esp_log.h"
+
+#include <etl/string.h>
+
+class Filesystem final : public sdk::Component {
+public:
+    Filesystem()  = default;
+    ~Filesystem() = default;
+
+    /* Component override functions */
+    etl::string<50> getTag() override { return TAG; };
+    Status          getStatus() override;
+    Status          initialize() override;
+    Status          run() override;
+    Status          stop() override;
+
+    // First is the size of the partition in bytes, second is the amount of used space in bytes
+    using PartitionInfo = std::pair<std::size_t, std::size_t>;
+
+    /**
+     * Write a file to the filesystem
+     * @tparam FILEPATH_SIZE Length of the filepath
+     * @tparam FILESIZE Length of the data to write
+     * @param filepath The path to the file to write
+     * @param data The data to write to the file
+     * @return std::error_code
+     */
+    template<std::size_t FILEPATH_SIZE, std::size_t FILESIZE>
+    static std::error_code writeFile(const etl::string<FILEPATH_SIZE>& filepath, const etl::string<FILESIZE>& data) noexcept {
+        const int fd = open(filepath.c_str(), O_WRONLY | O_CREAT, 0644);
+        if (fd == -1) { return std::error_code(errno, std::generic_category()); }
+
+        if (const auto ret = write(fd, data.c_str(), data.size()); ret == -1) {
+            close(fd);
+            return std::error_code(errno, std::generic_category());
+        }
+
+        close(fd);
+        return std::error_code();
+    }
+
+    /**
+     * Read a file from the filesystem
+     * @tparam FILEPATH_SIZE Length of the filepath
+     * @tparam FILESIZE Length of the data to read
+     * @param filepath The path to the file to read
+     * @return std::expected<etl::string<FILESIZE>, std::error_code>
+     */
+    template<std::size_t FILEPATH_SIZE, std::size_t FILESIZE>
+    static std::expected<etl::string<FILESIZE>, std::error_code> readFile(const etl::string<FILEPATH_SIZE>& filepath) noexcept {
```

Comment:
Might be cleaner to have the caller allocate a buffer and pass it to the `readFile` function with a size argument. 
Do check whether the passed array has the size mentioned by the argument :)